Add Pay in Person toggle on payments screen

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #7269 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds the `Pay in Person` option on the payments screen.

The `Pay in Person` option lets the user toggle the pay in person option. They can enable or disable it and the value will be reflected on the store admin page.

Along with this, this PR also divides existing payments screen options into 2 sections
1. Payment options: Collect Payment, Enable Pay in Person, Change Payment Provider
2. Card Readers: Order Card Reader, Manage Card Reader, Card Reader Manuals

Also, we have configured the list of options on the payment screen in such a way that items can be easily re-arranged in any desired position.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
## Enabling Pay in Person
1. Make sure `Pay in Person` is disabled on your store
2. Navigate to the Menu -> Payments screen
3. Enable `Pay in Person`
4. Go to the wp-admin page (settings -> Payments) of your site and ensure the `Cash on Delivery` setting is enabled.
5. Navigate back from the payments screen on the app and visit the payment screen again
6. Ensure the `Pay in Person` setting is enabled this time.

## Disabling Pay in Person
1. Make sure `Pay in Person` is enabled on your store
2. Navigate to the Menu -> Payments screen
3. Disable `Pay in Person`
4. Go to the wp-admin page (settings -> Payments) of your site and ensure the `Cash on Delivery` setting is disabled.
5. Navigate back from the payments screen on the app and visit the payment screen again
6. Ensure the `Pay in Person` setting is disabled this time.

## Enabling Pay in Person from the onboarding screen
1. Make sure `Pay in Person` is disabled on your store
2. Navigate to the Menu -> Payments screen
3. Tap on the onboarding error CTA at the bottom.
4. Tap on the "Enable Pay in Person" CTA
5. App redirects back to the payment screen after cod is enabled successfully
6. Make sure the `Pay in Person` setting is enabled now.
7. Go to the wp-admin page (settings -> Payments) of your site and ensure the `Cash on Delivery` setting is enabled.

## Pay in Person onboarding screen error
1. Make sure `Pay in Person` is disabled on your store
2. Navigate to the Menu -> Payments screen
3. Ensure you see onboarding not completed error.
4. Enable the `Pay in Person` option from the payment screen.
5. Navigate back from the payments screen and come back again.
6. Ensure you don't see any onboarding not completed error now.


## Payment screen sections
1. Navigate to the Menu -> Payments screen
2. Ensure the options are displayed under 2 sections properly
    1. Payment options: Collect Payment, Enable Pay in Person, Change Payment Provider
    2. Card Readers: Order Card Reader, Manage Card Reader, Card Reader Manuals

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/1331230/186848119-b3b378ea-37d4-4984-ad54-2f97e0074a40.mov



- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Add a new layout for toggleable options in payments

AnirudhBhat committed:
Add string resource for cod in payments screen

AnirudhBhat committed:
Make the hub screen support multiple view types.

In this case, we are talking about regular payment options (Non toggleable) and COD (Toggleable)

AnirudhBhat committed:
Add header type layout xml that will be shown in the hub screen.

AnirudhBhat committed:
Add string resource for the cod enable/disable option description

AnirudhBhat committed:
Add multiple types for the hub screen

This commit also contains the ability to sort the payment options before displaying. Sorting is configurable via the "index" parameter on the list item.

AnirudhBhat committed:
Extract cash on delivery toggler and querying logic into a separate class

AnirudhBhat committed:
Replace recycler adapter with list adapter to make use of diff util.

AnirudhBhat committed:
Refactor and add implementation for toggling cod from the payments screen

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat committed:
Merge branch 'trunk' into issue/7269-add-cod-toggle-in-payments-screen

AnirudhBhat committed:
Merge latest trunk and update CashOnDeliverySettings.kt

AnirudhBhat committed:
Add few more tests that verify correct labels and icons for cod state in payments screen

AnirudhBhat committed:
Added release notes

AnirudhBhat committed:
Fix a minor bug of UI not reflecting the latest state when cod option is toggled.

AnirudhBhat committed:
Merge branch 'trunk' into issue/7269-add-cod-toggle-in-payments-screen

AnirudhBhat committed:
Remove unnecessary code

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewModel.kt:
```diff
@@ -41,18 +44,55 @@ class CardReaderHubViewModel @Inject constructor(
 
+    private val cashOnDeliveryState = MutableLiveData(
+        ToggleableListItem(
+            icon = R.drawable.ic_manage_card_reader,
```
Need to change the icon for this item. Have asked [here](p1661494253973009/1660624954.229159-slack-C0354HSNUJH)

AnirudhBhat committed:
refactor cash on delivery state creation during view resume.

kidinov commented on RELEASE-NOTES.txt:
```diff
@@ -1,6 +1,7 @@
 *** PLEASE FOLLOW THIS FORMAT: [<priority indicator, more stars = higher priority>] <description> [<PR URL>]
 10.2
 -----
+-[**] In-Person Payments: Added Pay in Person toggle option in the Payments(Menu -> Payments) screen. Merchants can enable or disable "Pay in Person" from the app [https://github.com/woocommerce/woocommerce-android/pull/7280]
```
np: should it be `Payments (Meny ->` (space in between)

AnirudhBhat replied:
Agree. Fixed here: https://github.com/woocommerce/woocommerce-android/pull/7280/commits/9e110c60f06f7f94f0740de893f4c5fca3b77a5a

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CashOnDeliverySettings.kt:
```diff
@@ -0,0 +1,36 @@
+import javax.inject.Inject
+
+@Reusable
+class CashOnDeliverySettings @Inject constructor(
```
np: shell we call it `CashOnDeliverySettingsRepository` to be more consistent?

AnirudhBhat replied:
Good suggestion. Fixed here: https://github.com/woocommerce/woocommerce-android/pull/7280/commits/0d085cb0d09a3b25c95e9a268281124b901b9012

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CashOnDeliverySettings.kt:
```diff
@@ -0,0 +1,36 @@
+            site = selectedSite.get(),
+            gatewayId = GatewayRestClient.GatewayId.CASH_ON_DELIVERY,
+            enabled = shouldEnable,
+            title = "Pay in Person",
```
Have not been following the project in the details that's why asking just out of curiosity - is this exactly title will be set? What about localisation?

AnirudhBhat replied:
I played around with changing the store language on the admin page and saw that it had no effect on the title, description, and instruction of COD. Hence I decided to go with just English for now. I'll double-check this with the backend team though.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubAdapter.kt:
```diff
@@ -1,24 +1,83 @@
+                CardReaderHubViewHolder.HeaderViewHolder(parent)
+            }
+            else -> {
+                throw IllegalStateException("Unknown section")
```
np: detekt plugin advices to use `error("Unknown section")` instead of throwing `IllegalStateException`. Maybe consider using this?

AnirudhBhat replied:
Fixed here: https://github.com/woocommerce/woocommerce-android/pull/7280/commits/f6a3f486a7b89595398f9c8c27301715ddcb0b4a

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubAdapter.kt:
```diff
@@ -1,24 +1,83 @@
+        override fun areItemsTheSame(
+            oldItem: CardReaderHubViewModel.CardReaderHubViewState.ListItem,
+            newItem: CardReaderHubViewModel.CardReaderHubViewState.ListItem
+        ): Boolean {
```
np: would it make it easier if we override `equals` and `hashCode` on the `HeaderItem`, `ToggleableListItem`, `NonToggleableListItem` and then will return from here just `oldItem == newItem`?

In this case we should change `areContentsTheSame` though

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewHolder.kt:
```diff
@@ -31,4 +34,34 @@ sealed class CardReaderHubViewHolder(val parent: ViewGroup, @LayoutRes layout: I
+        CardReaderHubViewHolder(parent, R.layout.card_reader_hub_toggelable_item) {
+        var binding: CardReaderHubToggelableItemBinding = CardReaderHubToggelableItemBinding.bind(itemView)
+        override fun onBind(uiState: CardReaderHubViewModel.CardReaderHubViewState.ListItem) {
+            uiState as CardReaderHubViewModel.CardReaderHubViewState.ListItem.ToggleableListItem
```
np: Do we need to the ViewHolders as a sealed class if we anyway do the manual casting? Can `CardReaderHubViewHolder` be a simple abstract class?

AnirudhBhat replied:
Agree. Fixed here: https://github.com/woocommerce/woocommerce-android/pull/7280/commits/c9bb77f7129ef3396334168a5aee584aa142907b

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewModel.kt:
```diff
@@ -173,6 +252,28 @@ class CardReaderHubViewModel @Inject constructor(
+                cashOnDeliveryState.value?.copy(isEnabled = false, isChecked = isChecked)!!
+            )
+            val result = cashOnDeliveryToggler.toggleCashOnDeliveryOption(isChecked)
+            result.model?.let {
```
`WooResult` is returned here. Shell we determine if the call was successful using `isError`? I am just thinking about the case when an error happened but the model is not null

AnirudhBhat replied:
Good point! Fixed here: https://github.com/woocommerce/woocommerce-android/pull/7280/commits/28600b667440ba96314d7f6c90bc854fa6e9963c

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewHolder.kt:
```diff
@@ -31,4 +34,34 @@ sealed class CardReaderHubViewHolder(val parent: ViewGroup, @LayoutRes layout: I
+            binding.cardReaderHubSwitch.setOnCheckedChangeListener(null)
+            binding.cardReaderHubSwitch.isChecked = uiState.isChecked
+            binding.cardReaderHubSwitch.isClickable = uiState.isEnabled
+            binding.cardReaderHubSwitch.setOnCheckedChangeListener { _, isChecked ->
```
Shell we make the whole card clickable, so a user won't need to click only on the toggle? I think this is more like a standard android experience

AnirudhBhat replied:
Agree. Fixed here: https://github.com/woocommerce/woocommerce-android/pull/7280/commits/71ba50b9a34409a3cd02af4b7f5019c276251369

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewModel.kt:
```diff
@@ -173,6 +252,28 @@ class CardReaderHubViewModel @Inject constructor(
+                    cashOnDeliveryState.value?.copy(isEnabled = true, isChecked = isChecked)!!
+                )
+            } ?: run {
+                cardReaderTracker.trackCashOnDeliveryEnabledFailure(
```
Would it make sense to show at least some toast message indicating that error happened? Otherwise now the toggle goes back and it's not clear what's going on

AnirudhBhat replied:
Ahh! 🤦 I decided to add it later during the development and completely forgot about it later. Thanks for catching this!
https://github.com/woocommerce/woocommerce-android/pull/7280/commits/c5551a5d2f98131f8f5e355bd255d33ca6b95f28

kidinov commented:
If I disable the toggle and still go to the payment method selection screen, I still see `card` and `cash` options. While doing the same but via the "collect payment" button from the order it probably will make sense, so not sure what will be the best option here

kidinov commented:
Might be a bug:
* Disable the toggle
* Go and try to collect a card payment
* On the first attempt it shows the error screen with skip button
* I click the skip button and now everytime I am trying to collect a payment even with disabled toggle it just goes normally

https://user-images.githubusercontent.com/4923871/187195218-ad512d1e-9013-4812-bad5-b7a8f9c05c4d.mp4

kidinov reviewed (commented):
@AnirudhBhat Love the changes and making this screen easier to enhance! I left some comments, please take a look

pachlava commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -972,6 +972,8 @@
         -->
     <string name="payments_hub_title">Payments</string>
     <string name="card_reader_manage_card_reader">Manage Card Reader</string>
+    <string name="card_reader_enable_pay_in_person">Enable Pay in Person</string>
```
Since this is a label for the toggle switch, I think it's self-explanatory that depending on the switch state, `Pay in Person` will either be enabled or disabled. Because of this, the word `Enable` here looks a bit redundant, IMO:

```suggestion
    <string name="card_reader_enable_pay_in_person">Pay in Person</string>
```

AnirudhBhat replied:
Thanks for the suggestion @pachlava I agree that the word "Enable" doesn't make sense. Removed here: https://github.com/woocommerce/woocommerce-android/pull/7280/commits/ca1cec047f4fcfae856bfc8b0d70c06fc31d1620

AnirudhBhat commented:
>Might be a bug:

>Disable the toggle
Go and try to collect a card payment
On the first attempt it shows the error screen with skip button
I click the skip button and now everytime I am trying to collect a payment even with disabled toggle it just goes normally

This is the expected behavior. We felt that showing `COD disabled` screen everytime during IPP onboarding was too much since enabling COD isn't a requirement for IPP to work. We decided that tapping on the skip button will not show "COD disabled" screen again unless the user logs out or clears app data. 

You can read about this decision here: pdfdoF-167-p2#comment-2210

AnirudhBhat committed:
Add space between brace and word in the RELEASE-NOTES.txt

AnirudhBhat committed:
rename CashOnDeliverySettings.kt to CashOnDeliverySettingsRepository.kt

AnirudhBhat committed:
Fix Detekt complain

Replace throw IllegalStateException("Unknown section") with throw error("Unknown section") to make Detekt happy

AnirudhBhat committed:
refactor COD network call handling in CardReaderHubViewModel.kt

AnirudhBhat committed:
Make CardReaderHubViewHolder.kt abstract class

AnirudhBhat committed:
Make the entire enable pay in person card toggleable.

AnirudhBhat committed:
Show a toast message when toggling pay in person fails

AnirudhBhat committed:
Merge branch 'trunk' into issue/7269-add-cod-toggle-in-payments-screen

AnirudhBhat committed:
Disable toggle when network call is in progress

AnirudhBhat committed:
Fix detekt indentation errors

AnirudhBhat committed:
Add icon for Pay in Person row in the hub screen

AnirudhBhat committed:
Remove "Enable" word from "Enable Pay in Person" for the toggle.
